src/solc_0.7/GameToken/GameToken.sol

```diff
@@ -0,0 +1,380 @@
+    }
+
+    /// @notice Get the creator of the token type `id`.
+    /// @param id the id of the token to get the creator of.
+    /// @return the creator of the token type `id`.
+    function creatorOf(uint256 id) external view override returns (address) {
+        require(id != uint256(0), "NEVER_MINTED");
+        address originalCreator = address(id / CREATOR_OFFSET_MULTIPLIER);
+        address newCreator = _creatorship[originalCreator];
+        if (newCreator != address(0)) {
+            return newCreator;
+        }
+        return originalCreator;
+    }
+
+    /// @notice Function to get all assets and their quantities for a GAME
+    /// @param gameId The id of the GAME to get assets for
+    function getGameAssets(uint256 gameId) external view override returns (uint256[] memory, uint256[] memory) {
+        uint256 assetLength = _gameData[gameId]._assets.length();
+        uint256[] memory gameAssets;
+        uint256[] memory quantities;
+
+        if (assetLength != 0) {
+            gameAssets = new uint256[](assetLength);
+            quantities = new uint256[](assetLength);
+            for (uint256 i = 0; i < assetLength; i++) {
+                gameAssets[i] = _gameData[gameId]._assets.at(i);
+                quantities[i] = _gameData[gameId]._values[gameAssets[i]];
+            }
+        } else {
+            gameAssets = new uint256[](1);
+            quantities = new uint256[](1);
+            gameAssets[0] = uint256(0);
+            quantities[0] = uint256(0);
+        }
+        return (gameAssets, quantities);
+    }
+
+    /// @notice return the current minter
+    /// @return address of minter
+    function getMinter() external view override returns (address) {
+        return _minter;
+    }
+
+    function onERC1155BatchReceived(
+        address operator,
+        address from,
+        uint256[] calldata ids,
+        uint256[] calldata values,
+        bytes calldata data
+    ) external view override returns (bytes4) {
+        require(msg.sender == address(_asset), "UNAUTHORIZED_SENDER");
+        // @review
+        // require(from == operator, "SELF_EXECUTED_TRANSFER_ONLY");
+        require(ids.length > 0, "EMPTY_TRANSFER_DISALLOWED");
+        require(ids.length == values.length, "need to contains Asset");
+        // @review
+        // require(data.length > 0, "EMPTY_DATA_DISALLOWED");
+        return ERC1155_BATCH_RECEIVED;
+    }
+
+    function onERC1155Received(
+        address operator,
+        address from,
+        uint256 id,
+        uint256 value,
+        bytes calldata data
+    ) external view override returns (bytes4) {
+        require(msg.sender == address(_asset), "UNAUTHORIZED_SENDER");
+        // @review below.
+        // require(from == operator, "SELF_EXECUTED_TRANSFER_ONLY");
+        require(id != uint256(0) && value > 0, "EMPTY_TRANSFER_DISALLOWED");
+        // @review
+        // require(data.length > 0, "EMPTY_DATA_DISALLOWED");
+        return ERC1155_RECEIVED;
+    }
```

Comment:
Originally I had these reverting automatically. But after thinking about it more, I realized that the GameToken contract should to be able to receive erc1155 assets safely.
I've used `BundleSandSale.sol` as an example, but still needs some fine-tuning I think.
cc @wighawag  Thoughts?

Reply:
Yes this is an option, the issue is that unless the user send extra data (in the data field) you cannot know to what game token the assets are supposed to be added
so I think only be able to add through special function is more explicit

Reply:
I can see the concern around needing the extra data...
I was thinking along the lines of, limiting the callers of this function to the asset contract with :
`require(msg.sender == address(_asset), "UNAUTHORIZED_SENDER");` combined with the id param(which is the `gameId` which was created by the GameToken contract and passes to the asset contract in the all to `safeTransferFrom` .
But perhaps it's simpler to not use the asset's safeTransferFunctions. We know the GameToken contract knows how to deal with assets being transferred.
cc @wighawag

Reply:
I see in your comment above the suggestion to use:
```
if (operator == address(this)) { 
     return ERC1155_RECEIVED; 
 } 
 revert("ERC1155_REJECTED");
```
That's straightforward enough.

Reply:
done.